### PR TITLE
Run clippy on CI

### LINKED ISSUE #1402: Consider adding clippy on CI
Right now clippy(stable branch) on `winit` (X11/Wayland backends) produces 1500 lines of warnings/errors, so I think it makes sense to at least take a look on them, since most of the time they actually make sense.

### PR DESCRIPTION
**Pull request by kchibisov**
Fixes #1402.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -93,6 +96,10 @@ jobs:
         !contains(matrix.platform.target, 'wasm32'))
       run: cargo $CMD test --verbose --target ${{ matrix.platform.target }} $OPTIONS --features $FEATURES
 
+    - name: Lint with clippy
+      shell: bash
+      if: matrix.rust_version != 'nightly'
+      run: echo $FEATURES && cargo clippy --all-targets --target ${{ matrix.platform.target }} $OPTIONS --features $FEATURES
```
Comment (MarijnS95) [inline .github/workflows/ci.yml:102]: Do we need to `echo` what features are available? That should already be part of the `job` title because this is part of the `strategy: matrix:`?

Comment (kchibisov) [reply .github/workflows/ci.yml:102]: I just added it to test, since I want to understand how to blacklist the run on non-default-features.

File: src/platform_impl/ios/window.rs [hunk 2]
```diff
@@ -452,7 +452,7 @@ impl Window {
             // Like the Windows and macOS backends, we send a `ScaleFactorChanged` and `Resized`
             // event on window creation if the DPI factor != 1.0
             let scale_factor: CGFloat = msg_send![view, contentScaleFactor];
-            let scale_factor: f64 = scale_factor.into();
+            let scale_factor: f64 = scale_factor as _;
```
Comment (filnet) [inline src/platform_impl/ios/window.rs:455]: The variable type was removed in a similar expression higher up.

File: src/platform_impl/linux/wayland/window/shim.rs [hunk 3]
```diff
@@ -406,7 +406,8 @@ pub fn handle_window_requests(winit_state: &mut WinitState) {
     // Process the rest of the events.
     for (window_id, window_handle) in window_map.iter_mut() {
         let mut requests = window_handle.pending_window_requests.lock().unwrap();
-        for request in requests.drain(..) {
+        let requests = requests.drain(..);
+        for request in requests {
```
Comment (filnet) [inline src/platform_impl/linux/wayland/window/shim.rs:410]: is that change necessary ?

Comment (kchibisov) [reply src/platform_impl/linux/wayland/window/shim.rs:410]: Sort of. clippy on nightly suggested it.

File: src/platform_impl/linux/x11/util/geometry.rs [hunk 4]
```diff
@@ -370,16 +370,20 @@ impl XConnection {
...
+            );
             FrameExtentsHeuristic {
                 frame_extents,
                 heuristic_path: UnsupportedNested,
             }
         } else {
             // This is the case for xmonad and dwm, AKA the only WMs tested that supplied a
             // border value. This is convenient, since we can use it to get an accurate frame.
-            let frame_extents = FrameExtents::from_border(border.into());
+            let frame_extents = FrameExtents::from_border(border as c_ulong);
```
Comment (filnet) [inline src/platform_impl/linux/x11/util/geometry.rs:386]: Which clippy gets triggered here ?

PS: the best is to do one commit per clippy type...

File: src/platform_impl/macos/util/async.rs [hunk 5]
```diff
@@ -170,7 +170,6 @@ pub unsafe fn set_maximized_async(
             let curr_mask = ns_window.styleMask();
             if shared_state_lock.fullscreen.is_some() {
                 // Handle it in window_did_exit_fullscreen
-                return;
```
Comment (filnet) [inline src/platform_impl/macos/util/async.rs:173]: probably some clippy but scary...

File: src/platform_impl/ios/window.rs (hunk 2 above)
Comment (kchibisov) [reply src/platform_impl/ios/window.rs:455]: I don't understand what do you want from me here. The problem is that CGFLoat is either f32 or f64, but winit API is f64, so we need to cast it to f64 here. Using `.into()` won't work, since clippy finds it as redundant when `CGFloat` is `f64`, however we can cast it like `as _` and explicitly state that we assign to `f64` value.

File: src/platform_impl/macos/util/async.rs (hunk 5 above)
Comment (kchibisov) [reply src/platform_impl/macos/util/async.rs:173]: Yeah, will refactor it.

**Comment by kchibisov** [comment]
@MarijnS95 Do you happen to know why android target randomly fails from time to time? I think I've seen it a bunch of times here, rerunning do fix the issues.

File: src/event_loop.rs [hunk 6]
```diff
@@ -221,6 +223,12 @@ impl EventLoop<()> {
     }
 }
 
+impl Default for EventLoop<()> {
```
Comment (madsmtm) [inline src/event_loop.rs:226]: Nit: Add changelog entry

File: src/lib.rs [hunk 7]
```diff
@@ -132,6 +132,9 @@
 
 #![deny(rust_2018_idioms)]
 #![deny(rustdoc::broken_intra_doc_links)]
+#![deny(clippy::all)]
+#![cfg_attr(feature = "cargo-clippy", deny(warnings))]
```
Comment (madsmtm) [inline src/lib.rs:136]: How does this work?

File: examples/fullscreen.rs [hunk 8]
```diff
@@ -34,8 +34,8 @@ fn main() {
     event_loop.run(move |event, _, control_flow| {
         control_flow.set_wait();
 
-        match event {
-            Event::WindowEvent { event, .. } => match event {
+        if let Event::WindowEvent { event, .. } = event {
+            match event {
```
Comment (madsmtm) [inline examples/fullscreen.rs:38]: See [my comment on the last Clippy PR](https://github.com/rust-windowing/winit/pull/2108#pullrequestreview-839703098), I still think there is value in making the examples use `match` instead of `if let`.

File: src/platform_impl/macos/view.rs [hunk 9]
```diff
@@ -901,7 +897,7 @@ extern "C" fn insert_tab(this: &Object, _sel: Sel, _sender: id) {
         let first_responder: id = msg_send![window, firstResponder];
         let this_ptr = this as *const _ as *mut _;
         if first_responder == this_ptr {
-            let (): _ = msg_send![window, selectNextKeyView: this];
+            let _: () = msg_send![window, selectNextKeyView: this];
```
Comment (madsmtm) [inline src/platform_impl/macos/view.rs:900]: Haha, nice!

File: src/window.rs [hunk 10]
```diff
@@ -1126,15 +1126,15 @@ impl Default for CursorIcon {
 }
 
 /// Fullscreen modes.
-#[derive(Clone, Debug, PartialEq)]
+#[derive(Clone, Debug, PartialEq, Eq)]
```
Comment (madsmtm) [inline src/window.rs:1129]: Nit: Add changelog entry

**Review by madsmtm — CHANGES_REQUESTED** [review]
iOS and macOS changes looks fine (I never did like the `let () = msg_send![...]`, so cool that Clippy doesn't either).

File: src/platform_impl/macos/window_delegate.rs [hunk 11]
```diff
@@ -269,7 +269,7 @@ extern "C" fn window_will_close(this: &Object, _: Sel, _: id) {
         autoreleasepool(|| {
             // Since El Capitan, we need to be careful that delegate methods can't
             // be called after the window closes.
-            let () = msg_send![*state.ns_window, setDelegate: nil];
+            let _: () = msg_send![*state.ns_window, setDelegate: nil];
```
[the remark this answers is not in the document]
Comment (MarijnS95) [reply src/platform_impl/macos/window_delegate.rs:272]: Ah I see now that you added `allow(clippy::let_unit_value)` for this, which is why my `clippy` is happy about it. Then you don't need to make this change anymore too :)

File: examples/request_redraw_threaded.rs [hunk 12]
```diff
@@ -28,10 +28,9 @@ fn main() {
         control_flow.set_wait();
 
         match event {
-            Event::WindowEvent { event, .. } => match event {
-                WindowEvent::CloseRequested => control_flow.set_exit(),
-                _ => (),
-            },
+            Event::WindowEvent { event, .. } if event == WindowEvent::CloseRequested => {
```
Comment (MarijnS95) [inline examples/request_redraw_threaded.rs:31]: It seems common to handle these as:
```suggestion
            Event::WindowEvent { event: WindowEvent::CloseRequested, .. } => {
```

File: src/lib.rs (hunk 7 above)
Comment (MarijnS95) [reply src/lib.rs:136]: `clippy` enables the `cargo-clippy` feature on all crates when running through it, automatically denying all `rustc` warnings in this instance. However, this implies warnings from `clippy::all` (that lint group should only include warn/deny lints) and the `#![deny(clippy::all)]` is thus superfluous.

File: src/platform_impl/android/mod.rs [hunk 13]
```diff
@@ -845,18 +845,17 @@ impl MonitorHandle {
 
     pub fn video_modes(&self) -> impl Iterator<Item = monitor::VideoMode> {
         let size = self.size().into();
-        let mut v = Vec::new();
         // FIXME this is not the real refresh rate
         // (it is guarunteed to support 32 bit color though)
```
Comment (MarijnS95) [inline src/platform_impl/android/mod.rs:849]: Can you fix the typo here as a drive-by fix :)

File: src/platform_impl/ios/mod.rs [hunk 14]
```diff
@@ -109,8 +110,6 @@ pub enum OsError {}
 
 impl fmt::Display for OsError {
     fn fmt(&self, _: &mut fmt::Formatter<'_>) -> fmt::Result {
-        match self {
-            _ => unreachable!(),
-        }
+        unreachable!()
```
Comment (MarijnS95) [inline src/platform_impl/ios/mod.rs:113]: `unreachable!()` typically means something the compiler hasn't been able to deduce... This type is `pub`, are we sure no-one is ever able to call `Display` on `OsError`?

File: src/platform_impl/linux/wayland/window/shim.rs (hunk 3 above)
Comment (kchibisov) [reply src/platform_impl/linux/wayland/window/shim.rs:410]: Yes. I don't quite get as well, but I'm not that against either.

File: src/lib.rs (hunk 7 above)
Comment (kchibisov) [reply src/lib.rs:136]: deny clippy all means that it'll error on clippy warnings instead of treating them as warnings.

File: .github/workflows/ci.yml [hunk 15]
```diff
@@ -93,6 +95,10 @@ jobs:
         !contains(matrix.platform.target, 'wasm32'))
       run: cargo $CMD test --verbose --target ${{ matrix.platform.target }} $OPTIONS --features $FEATURES
 
+    - name: Lint with clippy
+      shell: bash
+      if: (matrix.rust_version != 'nightly') && !contains(matrix.platform.options, '--no-default-features')
+      run: cargo clippy --all-targets --target ${{ matrix.platform.target }} $OPTIONS --features $FEATURES
```
[the remark this answers is not in the document]
Comment (kchibisov) [reply .github/workflows/ci.yml:101]: I think we don't want to run clippy for examples actually.

File: src/platform_impl/windows/ime.rs [hunk 16]
```diff
@@ -80,19 +80,18 @@ impl ImeContext {
...
-        let size = ImmGetCompositionStringW(self.himc, gcs_mode, null_mut(), 0);
-        if size < 0 {
-            return None;
-        } else if size == 0 {
-            return Some(Vec::new());
-        }
+        let size = match ImmGetCompositionStringW(self.himc, gcs_mode, null_mut(), 0) {
+            0 => return Some(Vec::new()),
+            size if size < 0 => return None,
+            size => size,
```
[the remark this answers is not in the document]
Comment (kchibisov) [reply src/platform_impl/windows/ime.rs:93]: No, it suggested to refactor into `match`, so did I.

File: src/platform_impl/web/web_sys/mod.rs [hunk 17]
```diff
@@ -98,7 +98,7 @@ pub fn set_canvas_style_property(raw: &HtmlCanvasElement, property: &str, value:
     let style = raw.style();
     style
         .set_property(property, value)
-        .expect(&format!("Failed to set {}", property));
+        .unwrap_or_else(|_| panic!("Failed to set {}", property))
```
[the remark this answers is not in the document]
Comment (kchibisov) [reply src/platform_impl/web/web_sys/mod.rs:101]: I just refactored the old code to be the same? At least I've autoapplied clippy suggestion here.

File: src/platform_impl/macos/window_delegate.rs [hunk 18]
```diff
@@ -550,12 +551,12 @@ extern "C" fn window_did_fail_to_enter_fullscreen(this: &Object, _: Sel, _: id)
             shared_state.target_fullscreen = None;
         });
         if state.initial_fullscreen {
-            let _: () = unsafe {
-                msg_send![*state.ns_window,
+            unsafe {
+                let _: () = msg_send![*state.ns_window,
```
[the remark this answers is not in the document]
Comment (kchibisov) [reply src/platform_impl/macos/window_delegate.rs:555]: I've manually updated all the `let ()`, since I hate them. And we have half of the code with `let _: ()` and `let ()`, the first one is more clear, since it clearly states that we bind the type.

File: src/platform_impl/linux/x11/util/randr.rs [hunk 19]
```diff
@@ -164,7 +164,8 @@ impl XConnection {
         (self.xrandr.XRRFreeOutputInfo)(output_info);
         Some((name, scale_factor, modes))
     }
-    pub fn set_crtc_config(&self, crtc_id: RRCrtc, mode_id: RRMode) -> Result<(), ()> {
+
+    pub fn set_crtc_config(&self, crtc_id: RRCrtc, mode_id: RRMode) -> Option<()> {
```
[the remark this answers is not in the document]
Comment (kchibisov) [reply src/platform_impl/linux/x11/util/randr.rs:168]: Yeah, should probably add one, the point is that this function call literally does nothing in return.

File: src/platform_impl/ios/mod.rs (hunk 14 above)
Comment (kchibisov) [reply src/platform_impl/ios/mod.rs:113]: Yeah, I'll add some stub.

File: src/platform_impl/windows/ime.rs (hunk 16 above)
[the remark this answers is not in the document]
Comment (kchibisov) [reply src/platform_impl/windows/ime.rs:93]: Though, we need a value, so we either bind it and match with cmp, or do what I'm doing.

File: src/platform_impl/web/web_sys/mod.rs (hunk 17 above)
[the remark this answers is not in the document]
Comment (kchibisov) [reply src/platform_impl/web/web_sys/mod.rs:101]: I've include it, but there's no display for it, and I don't know anything about web to impl it myself.

File: .github/workflows/ci.yml (hunk 15 above)
[the remark this answers is not in the document]
Comment (MarijnS95) [reply .github/workflows/ci.yml:101]: I think we should be consistent and enforce it everywhere, so that the examples are also clean-enough.

File: src/window.rs (hunk 10 above)
Comment (kchibisov) [reply src/window.rs:1129]: I don't think internal impl requires it? Also I'm not sure that we should add changelog notes for `Eq` impls, since they are not breaking the code and transparent.

File: src/platform_impl/linux/x11/util/randr.rs (hunk 19 above)
[the remark this answers is not in the document]
Comment (MarijnS95) [reply src/platform_impl/linux/x11/util/randr.rs:168]: It doesn't mean the same though, the point is that the caller is supposed to check if this function succeeded or not. Return the `status: i32` in the `Err()` case for my part.

File: src/platform_impl/web/web_sys/mod.rs (hunk 17 above)
[the remark this answers is not in the document]
Comment (MarijnS95) [reply src/platform_impl/web/web_sys/mod.rs:101]: > I've include it, but there's no display for it, and I don't know anything about web to impl it myself.

This comment didn't show up before a page reload :grimacing: - expect uses `Debug` not `Display` so you should be able to trivially `: {:?}` add that to the `panic!`.

File: .github/workflows/ci.yml (hunk 15 above)
[the remark this answers is not in the document]
Comment (MarijnS95) [reply .github/workflows/ci.yml:101]: In addition there's `--all-targets` here and you fixed+allowed clippy lints in the `examples/`, so this is only sensible?

**Review by MarijnS95 — APPROVED** [review]
It looks like all my comments were addressed minus some minor nits which we can continue to discuss, nice job!

File: .github/workflows/ci.yml (hunk 15 above)
[the remark this answers is not in the document]
Comment (kchibisov) [reply .github/workflows/ci.yml:101]: I'd add `-D warnings` on CI then, I guess.

Having deny in lib.rs is still easier to understand that you must fix those issues.

[the remark this answers is not in the document]
Comment (MarijnS95) [reply .github/workflows/ci.yml:101]: Yeah, leave the deny in `lib.rs`, that should hopefully decrease the amount of noise in PRs that would otherwise require lots of cleanup pushes to "make the CI happy".

File: src/lib.rs (hunk 7 above)
Comment (MarijnS95) [reply src/lib.rs:136]: But that's also what `#![cfg_attr(feature = "cargo-clippy", deny(warnings))]` does :)

Comment (kchibisov) [reply src/lib.rs:136]: I think those about rustc warnings, and not clippy onces? I think it was like that in the past, at least.

File: src/window.rs (hunk 10 above)
Comment (madsmtm) [reply src/window.rs:1129]: Internal impls don't matter.

It's not a breaking change, but in general I think all API changes should be documented in the changelog, including small ones like this. But I don't have a strong opinion on this, you're free to disagree.

File: src/platform_impl/ios/mod.rs [hunk 20]
```diff
@@ -108,9 +109,7 @@ unsafe impl Sync for DeviceId {}
 pub enum OsError {}
 
 impl fmt::Display for OsError {
-    fn fmt(&self, _: &mut fmt::Formatter<'_>) -> fmt::Result {
-        match self {
-            _ => unreachable!(),
-        }
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "os error.")
```
Comment (maroider) [inline src/platform_impl/ios/mod.rs:113]: nit: this is hardly a sentence, so the period seems misplaced
```suggestion
        write!(f, "os error")
```

File: src/platform_impl/web/web_sys/scaling.rs [hunk 21]
```diff
@@ -65,13 +65,13 @@ impl ScaleChangeDetectorInternal {
         );
         let mql = MediaQueryListHandle::new(&media_query, closure);
         if let Some(mql) = &mql {
-            assert_eq!(mql.mql().matches(), true);
+            assert!(!mql.mql().matches());
```
Comment (maroider) [inline src/platform_impl/web/web_sys/scaling.rs:68]: It *looks* like the logic got accidentally negated here.

File: src/platform_impl/web/web_sys/scaling.rs [hunk 22]
```diff
@@ -65,13 +65,13 @@ impl ScaleChangeDetectorInternal {
...
         }
         mql
     }
 
     fn handler(&mut self, event: MediaQueryListEvent) {
-        assert_eq!(event.matches(), false);
+        assert!(event.matches());
```
Comment (maroider) [inline src/platform_impl/web/web_sys/scaling.rs:74]: ditto

File: .github/workflows/ci.yml (hunk 15 above)
[the remark this answers is not in the document]
Comment (maroider) [reply .github/workflows/ci.yml]: I'm fairly certain that the general advice here is to _not_ deny warnings in `*.rs` files, and instead only do it in CI, as it might break compilation with future versions of `rustc` and/or `clippy`.

[the remark this answers is not in the document]
Comment (kchibisov) [reply .github/workflows/ci.yml]: That's the point, we don't deny warnings for rustc or any related command, just when running clippy. So there's no way the build will break, only clippy could, but it would only mean that CI is broken.

**Review by madsmtm — APPROVED** [review]

File: .github/workflows/ci.yml (hunk 15 above)
[the remark this answers is not in the document]
Comment (maroider) [reply .github/workflows/ci.yml]: Ah, I see. That's clever.

File: src/window.rs (hunk 10 above)
Comment (maroider) [reply src/window.rs:1129]: A changelog entry would indeed be nice.

**Review by maroider — APPROVED** [review]
I don't have any further blocking concerns, although it would be nice to have changelog entries for all the added trait impls on public types.

**Review by msiglreith — APPROVED** [review]

**Comment by kchibisov** [comment]
Android failures were unrelated and quite random recently.

File: src/platform_impl/macos/mod.rs [hunk 23]
```diff
@@ -1,4 +1,5 @@
 #![cfg(target_os = "macos")]
+#![allow(clippy::let_unit_value)]
```
Comment (madsmtm) [inline src/platform_impl/macos/mod.rs:2]: That we have to specify this is actually a regression, I've filed an issue about it here: https://github.com/rust-lang/rust-clippy/issues/8998.
